app/src/main/java/com/example/tasky/feature_authentication/data/repository/AuthRepositoryImpl.kt

```diff
@@ -0,0 +1,68 @@
+package com.example.tasky.feature_authentication.data.repository
+
+import android.content.SharedPreferences
+import com.example.tasky.feature_authentication.data.remote.TaskyAuthApi
+import com.example.tasky.feature_authentication.data.remote.auth.LoginRequest
+import com.example.tasky.feature_authentication.data.remote.auth.RegistrationRequest
+import com.example.tasky.feature_authentication.data.util.AuthResult
+import com.example.tasky.feature_authentication.domain.repository.AuthRepository
+import retrofit2.HttpException
+
+class AuthRepositoryImpl(
+    private val api: TaskyAuthApi,
+    private val prefs: SharedPreferences
+) : AuthRepository{
+    override suspend fun signUp(fullName: String, email: String, password: String): AuthResult {
+        return try {
+            api.signUp(RegistrationRequest(
+                fullName = fullName,
+                email = email,
+                password = password
+            ))
+            signIn(email, password)
+        } catch (e: HttpException) {
+            if(e.code() == 401){
+                AuthResult.Unauthorized(e.message)
+            } else {
+                AuthResult.Error(e.message)
+            }
+        } catch (e: Exception){
+            AuthResult.Error(e.message)
+        }
```

Comment:
Be careful with catching general exceptions inside suspend functions. You can break coroutine cancellation that way. If a coroutine is cancelled, the currently suspending function (here, it would be `api.signUp`) will throw a CancellationException. If your catch block eats that up, this exception won't be properly propagated up the coroutine scopes which can cause some coroutines to run longer than they should.
You can fix this by adding this:

```
if(e is CancellationException) throw e
```
or alternatively by only catching specific exceptions and not CancellationExceptions.

In this video I show the issue with that a bit more in detail (starting at 17:45):
https://www.youtube.com/watch?v=VWlwkqmTLHc